content-services/content-sources-backend
# Fixes 4822: rename domains with cs- prefix

## Summary

Changes template model to no longer contain RepositoryConfigurations directly, but instead TemplateRepositoryConfigurations, this was due to an observed bug with gorm.  Strangely this bug was not always reproducible, but it was observed that querying Templates with the default scope resulted in RepositoryConfigurations included that had been removed from the template. 

This moves all the jobs to one command  ./cmd/jobs/main.go   that takes the name of the job as an argument.

Deletes the repair_latest and cleanup_versions jobs

Adds a new job 'rename_domains' that renames the red hat domain to 'cs-redhat' if its not already that.  Renames all other domains to cs-PREVIOUS_NAME  if it doesn't already start with 'cs-'.  

The rename process involves:

1.  loading the old name from the db
2. updating the paths for all the templates in candlepin
3. updating the content overrides for all the templates in candlepin TODO
4. updating the domain name in pulp
5. updating our own domains table

5 is done last so that if the process errors at any point, it can be re-run from the start for that domain simply by re-running the job.

## Testing steps

1.  On main, revert this commit: https://github.com/content-services/content-sources-backend/commit/ebdbba4f3841284682633178ea8f779193bf93ad
2. run:
```
make repos-import-rhel9
go run cmd/external-repos/main.go nightly-jobs 
```
start the server and let the RHEL repos snapshot
3.  Create at least 1 custom repo with snapshotting enabled
4. Create a content template, assign rhel9 and custom repo(s)
5. follow https://github.com/content-services/content-sources-backend/blob/main/docs/register_client.md to register a client
6. confirm that the client can install content form the redhat repos and custom repos
7. rename the domains:  `go run cmd/jobs/main.go rename_domains`
8. On the client run  `subscription-manager refresh`   && `subscription-manager repo-override`  
9. verify that you can still install content from the red hat and custom repos
10. confirm that the templates urls now point to   cs-DOMAIN  & cs-redhat
11. confirm that the repo configs in the webUI for snapshots still work
12. If you still see the old domains in the redhat.repo file, then try re-registering
